libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -404,7 +412,7 @@ void CodeTransform::operator()(FunctionDefinition const& _function)
 				subTransform.deleteVariable(var);
 		}
 
-	if (!m_allowStackOpt || _function.returnVariables.empty())
+	if (!m_allowStackOpt)
 		subTransform.setupReturnVariablesAndFunctionExit();
```

Comment:
Why was this removed?

Reply:
``setupReturnVariablesAndFunctionExit()`` clears the "unused" slots produced by the function arguments, so calling it early breaks functions with too many arguments if they don't have return variables. I.e.
```
function f(a1, ..., a20) {
mstore(<slot>, a1) // This is the last use of a1 and a1 needs to be popped after this statement for things to work.
// So it needs the ``!returnVariablesAndFunctionExitAreSetup()`` code path in ``freeUnusedVariables``.
}
```

Basically it was a mistake to have added it here in the first place :-).

Reply:
And maybe some more context: actually reusing the stack slots of arguments tends to be detrimental - the arguments are deep on the stack, putting anything there causes more problems than it solves. So we only want to actually pop them, if their last use is at the very beginning of the function before any other declaration, basically just to make it possible to move the arguments to memory at the beginning of the function.
That would basically be *why* ``setupReturnVariablesAndFunctionExit`` clears those unused slots.